### PR TITLE
Shaderc-rs & Shaderc-sys split

### LINKED ISSUE #44: Use Precompiled Sources to Link Against
I'm identifying a course of action to use pre-compiled sources to make it easier on users and package maintainers who need the rust bindings to **above all build**.

We have the .hpp and .so's in precompiled packages 
https://www.archlinux.org/packages/extra/x86_64/shaderc/

Obviously Gentoo will include basically the entire build
https://packages.gentoo.org/packages/media-libs/shaderc

As for how to convince build.rs to optionally favor a system package for the rest of the makefile and to not check out (or just not build them), I will have to start getting up to speed on how the rust package is deciding what to link against and how to add options.

If a course of implementation is recommended I can PR the changes as well as test on my CI server's and assist in getting pre-built tooling into maintenance and actually maintain it.

### PR DESCRIPTION
**Pull request by knappador**
--snip-- #44  has almost all of the discussion leading up to the merge commit notes

fixes #44

### CODE REVIEW

**Comment by knappador** [comment]
I signed it!
@googlebot

File: shaderc-rs/src/lib.rs
Comment (antiagainst) [inline shaderc-rs/src/lib.rs:78]: Nit: Can we put this line before `use libc::`? It is introducing an alias and different from others. :)

**Comment by knappador** [comment]
@antiagainst  I'm not sure what's up on the Appveyor results.  I have little visibility outside of Arch Linux so far.

This PR should read fine as a series of independent commits.

[Re-implemented the `--no-defaults` behavior](https://github.com/e-nguyen/shaderc-rs/pull/1)
[Added more robust build options & behavior](https://github.com/e-nguyen/shaderc-rs/pull/2)

I primarily used [this heavily referenced guide to writing sys crates](https://kornel.ski/rust-sys-crate) to select what traditions to try to follow.  The one area I'm not yet 100% clear on is [overriding the static library location using a downstream project's `Cargo.toml`](https://doc.rust-lang.org/cargo/reference/build-scripts.html#the-links-manifest-key)

To quickly run through a thorough test of the Linux build paths on Arch linux:

```
cargo test
cargo clean
cargo test --features build-from-source
cp ../target/debug/../libshaderc_combined.a .
cargo clean
cargo test SHADERC_STATIC=./libshaderc_combined.a
cargo clean
cargo test --features check-passthrough  # will cause trigger old --no-defaults behavior on sys crate
# cp the library into the specified location
# re-run with --features check-passthrough
cargo clean
mkdir libs
cp /usr/lib/libshaderc_combined.a libs/
cp /usr/lib/libglslang.a libs/
cp /usr/lib/libSPIRV.a libs/
cp /usr/lib/libSPIRV-Tools.a libs/
cargo test SHADERC_STATIC=./libs/
```
Happy to punch some commits into working on Appveyor but I'll need some information.

**Comment by knappador** [comment]
Handling the merge issue.  Tests for shaderc-rs fail due to linking.  I'm incorporating 367a4ddbaaacf9590a68102db62aa0a2bc3e8953 as well, which won't apply at all.

Might punch some more history later to convert the delete to a move if possible.

**Comment by knappador** [comment]
Full test sequence fine on my machine, on Linux.  There was a regression in build.rs.  glslang being included statically broke compiling the tests.  I have no idea how I found that when I wrote down the critical hunk in #44

**Comment by knappador** [comment]
@antiagainst   looks like all three platforms have some level of support based on tests?  Time to party yet?  :package: :tada: 

- [ ] Let's remember to toss in a version bump right before merge.  I think "0.5" is what we settled on for #44 

Feel free to keep pinging me on build issues moving forward.  

I'm nominating this for merge and not blocked on anything.

**Comment by knappador** [comment]
I'm still thinking about necessary ways to pass through the new options to shaderc-sys.  The command line settings "just work."  I'm going to add in a patch to pass through the build-from-source feature to shaderc-rs since it's just easier than figuring out a nested dependency configuration for most people.

File: shaderc-sys/src/lib.rs
Comment (knappador) [inline shaderc-sys/src/lib.rs:46]: Just remembered this.  I don't have a clear explanation of why this was needed, just that it broke without the change.

File: README.md
Comment (antiagainst) [inline README.md]: Nit: Can we wrap these extra long lines into 80 characters?

File: shaderc-rs/Cargo.toml
Comment (antiagainst) [inline shaderc-rs/Cargo.toml]: Yep, let's bump this to 0.5.

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: Nit: I think it's more of `emit_std_cpp_link_lib` here.

File: shaderc-sys/src/lib.rs
Comment (antiagainst) [reply shaderc-sys/src/lib.rs:46]: Because right now this file is moved into a new crate. Previously it is in the shaderc-rs crate, which had this in `lib.rs`.

File: shaderc-rs/src/lib.rs
Comment (antiagainst) [inline shaderc-rs/src/lib.rs:1783]: Normally it would be nice to separate these formatting changes with others. But I won't bother this time to avoid incurring more work for you. :)

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [inline shaderc-sys/Cargo.toml]: What about "Shaderc system library FFI, building, and linking" here? "Raw bindings" is not quite distinguishable with "Rust bindings" in the above `Cargo.toml`.

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [inline shaderc-sys/Cargo.toml]: Sorry I don't fully understand here. What's the purpose of `check-passthrough`? It seems we can merge it into `dont-use-deprecated` to be just one feature?

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [inline shaderc-sys/Cargo.toml]: Can we give this a more descriptive name? `dont-use-deprecated` is quite general and it is applicable to all kinds of deprecated features. What about "use-sys-libs" or something similar?

File: shaderc-rs/Cargo.toml
Comment (antiagainst) [inline shaderc-rs/Cargo.toml]: What about "prefer-sys-libs" or something similar to be more descriptive?

File: shaderc-rs/Cargo.toml
Comment (antiagainst) [inline shaderc-rs/Cargo.toml]: I'd suggest to make the comment here to be more clear:

```
# Disabling defaults to the deprecated behavior: not building from source,
not using system libraries, only relying on the developer to provide native
shaderc libraries
```

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: Should we have a check to error out if both `CARGO_FEATURE_DONT_USE_DEPRECATED` and `CARGO_FEATURE_BUILD_FROM_SOURCE` is true?

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: `--no-deafults` is for shaderc-rs. Here we are in shaderc-sys now. :) What about just saying `USE OF 'dont-use-deprecated' IS ...` ?

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: `Requested to skip finding system libraries or building native C++ shaderc` ?

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: Can we make this message a little bit more explicit by attaching the dir to it: `Using specified pre-built libshaderc in {}` ? It could be quite useful if something goes wrong regarding the directory so that the developer can easily notice it. :)

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: Question: Why are we having three environment variables for the same functionality here?

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: Should we prefer dynamic lib to static lib? What do you think? But it's fine to do in another PR.

File: shaderc-sys/build/build.rs
Comment (antiagainst) [inline shaderc-sys/build/build.rs]: I think we can do `Some((&search_dir, COMBINED_LIB))` and `Some((&search_dir, dylib_name.as_str()))` to avoid copying?

File: README.md
Comment (antiagainst) [inline README.md]: I think the explanation here is not very clear. It would be nice to document features, env vars, and lib resolution orders more clearly, like:

* The default behavior is to search and link system libraries for shaderc under the directory given by env var .... This first tries static library, then dynamic library. If none is found, then ...
* If `--no-default-features` is given, then ... This is deprecated ...
* If `build-from-source` is given, then ...

File: README.md
Comment (antiagainst) [inline README.md]: Could you also update the documentation in `lib.rs`? Thanks!

**Review by antiagainst — COMMENTED** [review]
This is awesome progress! :) I've a few comments inlined. Thanks for the work again!

File: shaderc-sys/build/build.rs
Comment (knappador) [reply shaderc-sys/build/build.rs]: I'm going to change it to only use `SHADERC_LIB_DIR` since we might have multiple libraries and calling it a path would be false.

A lot of the choices attempted to defer to this document: https://kornel.ski/rust-sys-crate

File: shaderc-sys/build/build.rs
Comment (knappador) [reply shaderc-sys/build/build.rs]: My vantage point is from years of only maintaining Linux machines, so it's actually quite hard for me to anticipate or reason about their choices.

I cargo-culted the bits that try to build a Windows dll path from the same document I linked before.  I truly have no idea what kind of library might get built on Windows.

File: shaderc-sys/Cargo.toml
Comment (knappador) [reply shaderc-sys/Cargo.toml]: I agree the name choices are not much use.  

As for the critical decision, using two features, it began with making shaderc-sys independent of shaderc-rs. 

Fundamentally, we are trying to re-export a feature of shaderc-sys's build options that begins by disabling the default features on shaderc-rs.  The only way to accomplish this is by inverting the feature.  This inversion also means any other package that doesn't set the feature would get `--no-defaults-behavior`.  Therefore, I protected this inverted feature with a second feature.

File: shaderc-sys/build/build.rs
Comment (knappador) [reply shaderc-sys/build/build.rs]: You might have been asking about dynamic on Linux.  I think I read that static was more likely to succeed.  Not specifying would be cool since it will search more accurately.

The pkg-config file has landed, but it will be a little bit before it hits package managers.  As for from source builds, I believe static is default when you bring your own version because you can only share if everyone is on the same page (using a system package manager) 
https://github.com/google/shaderc/commit/439a848c0c0e91b5f1b85059cd4b059274b9a93b

**Comment by knappador** [comment]
I didn't have a lot of time to review changes in https://github.com/google/shaderc-rs/pull/46/commits/f03ba636d9fa79b26c636269c0da01526a1e5ffa

Many of the comments were addressed.  I'll punch the revision history after changes are more final.

Worth considering how the new structure might affect the docs.  I have little experience yet working with Rust docs.
https://docs.rs/shaderc/0.3.16/shaderc/

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [reply shaderc-sys/Cargo.toml]: Okay, I see, thanks!

File: shaderc-sys/build/build.rs
Comment (antiagainst) [reply shaderc-sys/build/build.rs]: SG! I don't have strong opinions here. As you said, it's hard to have a satisfying default anyway. :)

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [inline shaderc-sys/Cargo.toml:23]: How do you think about `shaderc-rs-use-sys-libs` and `use-sys-libs` here then? Given the tight relationship between shaderc-rs and shaderc-sys, it is reasonable to have an option dedicated to shaderc-rs here. (I'm just trying to make the feature more self-documenting. :) )

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [reply shaderc-sys/Cargo.toml]: I think another way is just to eliminate the `--no-defaults-behavior` support for shaderc-rs entirely. It was not a satisfying solution; and now we have `SHADERC_LIB_DIR` now and it is better. WDYT?

**Review by antiagainst — APPROVED** [review]
This is nice, thanks again for doing this! :)

Feel free to do whatever clean up you'd like to and then let me know. I will apply a few cosmetic fixes and then merge this.

File: shaderc-sys/Cargo.toml
Comment (knappador) [reply shaderc-sys/Cargo.toml:23]: This crazy inversion thing is only necessary because of how `--no-default-features` operates on a crate's config.  I would support a rename to `shaderc` specific if you keep the backwards compatibility.  If you axe the backwards compatibility, these features go away.  `shaderc-rs-inverted-no-defaults` etc would be more appropriate.  The features have no control over reading specific directories, only the old behavior of reading the `../out/` directory.

If you decide to axe no defaults, it's as simple as removing these two features in both crates and axing the block in build.rs that reads them, then removing the documents.

The reason I implemented the `--no-default-features` backwards compatibility is because you believed people were using it (possibly in CI?).  It's wise to maintain at least one release of deprecation in any case.

File: shaderc-sys/build/build.rs
Comment (knappador) [reply shaderc-sys/build/build.rs]: (could be old question) Let's back up.  The way I preserved the old behavior is so that in `shaderc-rs` you can still say:
```$ cargo test --no-default-features``` 
And it just works like it does in master.

This is done by adding the two goofy features.  I don't anticipate that anyone will explicitly do something like:
```toml
[dependencies]
shaderc-sys = {version = "0.5.0", features = ["enable-inverted-no-defaults"]
```

Basically it's definitely a shaderc-rs choice, part-and-parcel with using shaderc-rs, but has to be implemented in shaderc-sys do to the way `--no-default-features` behaves

File: shaderc-sys/build/build.rs
Comment (knappador) [reply shaderc-sys/build/build.rs]: A lot of the options are indeed contradictory.  Stating the order of preference in the docs is sufficient IMO.

File: shaderc-rs/src/lib.rs
Comment (knappador) [reply shaderc-rs/src/lib.rs:78]: I believe this got fixed, but was not marked outdated somehow

https://github.com/google/shaderc-rs/blob/f03ba636d9fa79b26c636269c0da01526a1e5ffa/shaderc-rs/src/lib.rs#L66-L79

**Comment by knappador** [comment]
I added some cleanups after doing some testing and sitting back to leisurely read the code.  One feature I actually changed was to try to canonicalize relative paths so that they would work... only to realize that the CWD changes during each recursion cargo makes to a dependency.  Specifying a relative path must be done relative to `shaderc-sys` crate. 

--rerun-after-changed and it's env variable counterpart deserve a look.  I think we have to combine both for our case, one to detect lib file changes and the other to detect env changes.  They have some funky behavior though.  I didn't try them.

 AFAIK the only outstanding question is to completely axe `--no-default-features` behavior or leave it deprecated for a few release cycles.


Tests
```shell
cargo test
cargo clean
cargo test --features build-from-source
cp ../target/debug/../libshaderc_combined.a ../
cargo clean
cargo test SHADERC_LIB_DIR=../
cargo clean
cargo test --no-default-features
# cp the library into the specified location
# re-run  test --no-default-features
cargo clean
mkdir ../libs
cp /usr/lib/libshaderc_combined.a ../libs/
cp /usr/lib/libglslang.a ../libs/
cp /usr/lib/libSPIRV.a ../libs/
cp /usr/lib/libSPIRV-Tools.a ../libs/
cargo test SHADERC_LIB_DIR=../libs/
```

File: shaderc-sys/Cargo.toml
Comment (antiagainst) [reply shaderc-sys/Cargo.toml:23]: SG. This will be removed soon, so I won't bother that much anymore. :)

**Comment by antiagainst** [comment]
Looks good! I just did some final touch over documentation. Will merge soon.

Good point about `--no-default-features` support. I think with semver, it should be fine to axe it entirely. (Sorry I didn't anticipate it as being this complicated, so I suggested we should support it earlier. We have users of it, see #36) But you already implemented it, let's be nice to keep it for another release. Thanks! :)

**Comment by knappador** [comment]
:+1:   Changes look good.  Thanks for the extensive review, feedback, and refinements

https://github.com/vulkano-rs/vulkano/issues/1183 crosslinking

**Comment by antiagainst** [comment]
Merged, thanks! @knappador, it seems I cannot publish shaderc-sys because I don't have the right access. Would you mind add me? Or If you want to do it, please feel free (use the top of the tree). :)

**Comment by knappador** [comment]
I believe I invited you already, I will resend when I get to my credentials

**Comment by antiagainst** [comment]
https://crates.io/me/pending-invites shows I've no pending owner invites.

**Comment by knappador** [comment]
@antiagainst 

Invite sent.  If there's no crate, the UI presents no option to manage owners :speak_no_evil: 

`cargo publish` from within the shaderc-sys crate failed first time.  I changed the README path back to relative, success.  I can PR this if you're unsure of what I did to get cargo to succeed.  

Also it means I burned another semver to give you an invite :speak_no_evil:  0.4.2 I guess :cat:

**Comment by antiagainst** [comment]
Nice, I've access now and published 0.5.0! \o/ Thanks!

Could you yank 0.4.1 now? :)

**Comment by knappador** [comment]
Yanked.  Going to merge this downstream for CI justice.

**Pull request merged**
